ipnsfs/dir.go

```diff
@@ -0,0 +1,343 @@
+	_, err := d.fs.dserv.Add(nd)
+	if err != nil {
+		return err
+	}
+
+	d.lock.Lock()
+	err = d.node.RemoveNodeLink(name)
+	if err != nil && err != dag.ErrNotFound {
+		d.lock.Unlock()
+		return err
+	}
+
+	err = d.node.AddNodeLinkClean(name, nd)
+	if err != nil {
+		d.lock.Unlock()
+		return err
+	}
+	d.lock.Unlock()
+
+	return d.parent.closeChild(d.name, d.node)
+}
+
+func (d *Directory) Type() NodeType {
+	return TDir
+}
+
+func (d *Directory) childFile(name string) (*File, error) {
+	fi, ok := d.files[name]
+	if ok {
+		return fi, nil
+	}
+
+	// search dag
+	for _, lnk := range d.node.Links {
+		if lnk.Name == name {
+			nd, err := lnk.GetNode(d.fs.dserv)
+			if err != nil {
+				return nil, err
+			}
+			i, err := ft.FromBytes(nd.Data)
+			if err != nil {
+				return nil, err
+			}
+
+			switch i.GetType() {
+			case ufspb.Data_Directory:
+				return nil, ErrIsDirectory
+			case ufspb.Data_File:
+				nfi, err := NewFile(name, nd, d, d.fs)
+				if err != nil {
+					return nil, err
+				}
+				d.files[name] = nfi
+				return nfi, nil
+			case ufspb.Data_Metadata:
+				return nil, ErrNotYetImplemented
+			default:
+				return nil, ErrInvalidChild
+			}
+		}
+	}
+	return nil, ErrNoSuch
+}
+
+func (d *Directory) childDir(name string) (*Directory, error) {
+	dir, ok := d.childDirs[name]
+	if ok {
+		return dir, nil
+	}
+
+	for _, lnk := range d.node.Links {
+		if lnk.Name == name {
+			nd, err := lnk.GetNode(d.fs.dserv)
+			if err != nil {
+				return nil, err
+			}
+			i, err := ft.FromBytes(nd.Data)
+			if err != nil {
+				return nil, err
+			}
+
+			switch i.GetType() {
+			case ufspb.Data_Directory:
+				ndir := NewDirectory(name, nd, d, d.fs)
+				d.childDirs[name] = ndir
+				return ndir, nil
+			case ufspb.Data_File:
+				return nil, fmt.Errorf("%s is not a directory", name)
+			case ufspb.Data_Metadata:
+				return nil, ErrNotYetImplemented
+			default:
+				return nil, ErrInvalidChild
+			}
+		}
+
+	}
+
+	return nil, ErrNoSuch
+}
+
+func (d *Directory) Child(name string) (FSNode, error) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	dir, err := d.childDir(name)
+	if err == nil {
+		return dir, nil
+	}
+	fi, err := d.childFile(name)
+	if err == nil {
+		return fi, nil
+	}
+
+	return nil, ErrNoSuch
+}
+
+func (d *Directory) List() []string {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	var out []string
+	for _, lnk := range d.node.Links {
+		out = append(out, lnk.Name)
+	}
+	return out
+}
+
+func (d *Directory) Mkdir(name string) (*Directory, error) {
+	d.lock.Lock()
+
+	_, err := d.childDir(name)
+	if err == nil {
+		d.lock.Unlock()
+		return nil, errors.New("directory by that name already exists")
+	}
+	_, err = d.childFile(name)
+	if err == nil {
+		d.lock.Unlock()
+		return nil, errors.New("file by that name already exists")
+	}
+
+	ndir := &dag.Node{Data: ft.FolderPBData()}
+	err = d.node.AddNodeLinkClean(name, ndir)
+	if err != nil {
+		d.lock.Unlock()
+		return nil, err
+	}
+	d.lock.Unlock()
```

Comment:
locking here is a bit sketch. do we have to unlock for `closeChild`? if so, maybe using func defer?

```
func() {
  d.lock.Lock()
  defer d.lock.Unlock()
  
  ... stuff
}()
```